README
NAME
    Devel::StrictMode - determine whether strict (but slow) tests should be
    enabled

SYNOPSIS
       package MyClass;
   
       use Moose;
       use Devel::StrictMode;
   
       has input_data => (
          is       => 'ro',
          isa      => STRICT ? "HashRef[ArrayRef[Str]]" : "HashRef",
          required => 1,
       );

DESCRIPTION
    This module provides you with a constant `STRICT` which you can use to
    determine whether additional strict (but slow) runtime tests are executed
    by your code.

    `STRICT` is true if any of the following environment variables have been
    set to true:

       PERL_STRICT
       EXTENDED_TESTING
       AUTHOR_TESTING
       RELEASE_TESTING

    `STRICT` is false otherwise.

    It is anticipated that you might set one or more of the above variables to
    true while running your test suite, but leave them all false in your
    production scenario.

    Although not exported by default, a constant `LAX` is also provided, which
    returns the opposite of `STRICT`.

  Using STRICT with Moose/Moo/Mouse attributes
    Type constraint checks (`isa`) are conducted at run time. Slow checks can
    slow down your constructor and accessors. As shown above, `STRICT` can be
    used to alternate between a slower by stricter type constraint check, and
    a faster but looser one.

    Don't try this if your attribute coerces. It will subtly break things.

  Using STRICT to perform assertions in function and method calls
    You may protect blocks of assertions with an `if (STRICT) { ... }`
    conditional to ensure that they only run in your testing environment.

       sub fibonacci
       {
          my $n = $_[0];
      
          if (STRICT)
          {
             die "expected exactly one argument"
                unless @_ == 1;
             die "expected argument to be a natural number"
                unless $n =~ /\A[0-9]+\z/;
          }
      
          $n < 2 ? $n : fibonacci($n-1)+fibonacci($n-2);
       }

    Because `STRICT` is a constant, the Perl compiler will completely optimize
    away the `if` block when running in your production environment.

  Using STRICT with pragmata
    Thanks to if it's easy to use `STRICT` to conditionally load pragmata.

       use Devel::StrictMode;

       use strict;
       use warnings STRICT ? qw(all) : qw(FATAL all);
   
       no if STRICT, "bareword::filehandles";
       no if STRICT, "autovivification";

    See also autovivification, bareword::filehandles, indirect,
    multidimensional, etc.

BUGS
    Please report any bugs to
    <http://rt.cpan.org/Dist/Display.html?Queue=Devel-StrictMode>.

SEE ALSO
    strictures.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2014 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
Devel-StrictMode
================

Created:      2014-05-10
Home page:    <https://metacpan.org/release/Devel-StrictMode>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=Devel-StrictMode>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.002	2014-05-10

 [ Documentation ]
 - Minor fix in pragmata example.

0.001	2014-05-10	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"       => "determine whether strict (but slow) tests should be enabled",
  "author"         => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config" => 0,
  "generated_by"   => "Dist::Inkt::Profile::TOBYINK version 0.015, CPAN::Meta::Converter version 2.120921",
  "keywords"       => [],
  "license"        => ["perl_5"],
  "meta-spec"      => {
                        url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                        version => 2,
                      },
  "name"           => "Devel-StrictMode",
  "no_index"       => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"        => {
                        configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                        develop   => { recommends => { "Dist::Inkt" => 0.001 } },
                        runtime   => { requires => { perl => 5.006 } },
                        test      => { requires => { "Test::More" => 0.47 } },
                      },
  "provides"       => {
                        "Devel::StrictMode" => { file => "lib/Devel/StrictMode.pm", version => 0.002 },
                      },
  "release_status" => "stable",
  "resources"      => {
                        bugtracker   => {
                                          web => "http://rt.cpan.org/Dist/Display.html?Queue=Devel-StrictMode",
                                        },
                        homepage     => "https://metacpan.org/release/Devel-StrictMode",
                        license      => ["http://dev.perl.org/licenses/"],
                        X_identifier => "http://purl.org/NET/cpan-uri/dist/Devel-StrictMode/project",
                      },
  "version"        => 0.002,
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	test       => { TESTS => "t/*.t" },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

